tao_compiler/mlir/disc/transforms/disc_algebra_simplifier.cc

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 The BladeDISC Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+
+#define DEBUG_TYPE "disc-algebra-simplifier"
+
+// This file implements some basic optimizations to do algebra simplification.
+
+namespace mlir {
+namespace disc_ral {
+namespace {
+
+// convert:
+//   mhlo.pow(x, const integer n)
+// to:
+//   %1 = x * 1
+//   %2 = x * %1
+//   ... ...
+//   %n = x * %n-1
+struct ExpandPowOp : public OpRewritePattern<mhlo::PowOp> {
+  using OpRewritePattern<mhlo::PowOp>::OpRewritePattern;
+
+  LogicalResult expandPowOp(mhlo::PowOp op, PatternRewriter& rewriter,
+                            int64_t exponential) const {
+    // TODO(disc): support non-positive exponential
+    if (exponential <= 0) return failure();
```

Comment:
Shall we limit the max exponential to avoid bad case?

Reply:
done!